lib/auth/grpcserver_test.go

```diff
@@ -0,0 +1,301 @@
+					require.Equal(t, totpRegisterChallenge.Algorithm, otp.AlgorithmSHA1.String())
+					code, err := totp.GenerateCodeCustom(totpRegisterChallenge.Secret, clock.Now(), totp.ValidateOpts{
+						Period:    uint(totpRegisterChallenge.Period),
+						Digits:    otp.Digits(totpRegisterChallenge.Digits),
+						Algorithm: otp.AlgorithmSHA1,
+					})
+					require.NoError(t, err)
+
+					totpSecrets["totp-dev"] = totpRegisterChallenge.Secret
+					return &proto.MFARegisterResponse{
+						Response: &proto.MFARegisterResponse_TOTP{TOTP: &proto.TOTPRegisterResponse{
+							Code: code,
+						}},
+					}
+				},
+				wantDev: func(t *testing.T) *types.MFADevice {
+					wantDev, err := services.NewTOTPDevice("totp-dev", totpSecrets["totp-dev"], clock.Now())
+					require.NoError(t, err)
+					return wantDev
+				},
+			},
+		},
+		{
+			desc: "add a U2F device",
+			opts: mfaAddTestOpts{
+				initReq: &proto.AddMFADeviceRequestInit{
+					DeviceName: "u2f-dev",
+					Type:       proto.AddMFADeviceRequestInit_U2F,
+				},
+				authHandler: func(t *testing.T, req *proto.MFAAuthenticateChallenge) *proto.MFAAuthenticateResponse {
+					// Respond to challenge using the existing TOTP device.
+					require.NotNil(t, req.TOTP)
+					code, err := totp.GenerateCode(totpSecrets["totp-dev"], clock.Now())
+					require.NoError(t, err)
+
+					return &proto.MFAAuthenticateResponse{Response: &proto.MFAAuthenticateResponse_TOTP{TOTP: &proto.TOTPResponse{
+						Code: code,
+					}}}
+				},
+				registerHandler: func(t *testing.T, req *proto.MFARegisterChallenge) *proto.MFARegisterResponse {
+					u2fRegisterChallenge := req.GetU2F()
+					require.NotEmpty(t, u2fRegisterChallenge)
+
+					mdev, err := mocku2f.Create()
+					require.NoError(t, err)
+					u2fDevices["u2f-dev"] = mdev
+					mresp, err := mdev.RegisterResponse(&u2f.RegisterChallenge{
+						Challenge: u2fRegisterChallenge.Challenge,
+						AppID:     u2fRegisterChallenge.AppID,
+					})
+					require.NoError(t, err)
+
+					return &proto.MFARegisterResponse{Response: &proto.MFARegisterResponse_U2F{U2F: &proto.U2FRegisterResponse{
+						RegistrationData: mresp.RegistrationData,
+						ClientData:       mresp.ClientData,
+					}}}
+				},
+				wantDev: func(t *testing.T) *types.MFADevice {
+					wantDev, err := u2f.NewDevice(
+						"u2f-dev",
+						&u2f.Registration{
+							KeyHandle: u2fDevices["u2f-dev"].KeyHandle,
+							PubKey:    u2fDevices["u2f-dev"].PrivateKey.PublicKey,
+						},
+						clock.Now(),
+					)
+					require.NoError(t, err)
+					return wantDev
+				},
+			},
+		},
+	}
+	for _, tt := range addTests {
+		t.Run(tt.desc, func(t *testing.T) {
+			testAddMFADevice(ctx, t, cl, tt.opts)
+			// Advance the time to roll TOTP tokens.
+			clock.Advance(30 * time.Second)
+		})
+	}
+
+	// Check that all new devices are registered.
+	resp, err = cl.GetMFADevices(ctx, &proto.GetMFADevicesRequest{})
+	require.NoError(t, err)
+	require.Len(t, resp.Devices, len(addTests))
```

Comment:
I know this is harder but you probably want to make sure that only the expected devices are registered (and not e.g. the same twice)